src/analysis/normed_space/spectrum.lean

```diff
@@ -0,0 +1,182 @@
+/-
+Copyright (c) 2021 Jireh Loreaux. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+import algebra.algebra.spectrum
+import analysis.calculus.deriv
+/-!
+# The spectrum of elements in complete normed algebra
+
+This file contains the basic theory for the resolvent and spectrum of a Banach algebra.
+
+## Main definitions
+
+* `spectral_radius`: supremum of `abs k` for all `k ∈ spectrum 𝕜 a`
+
+## Main statements
+
+* `is_open_resolvent_set`: the resolvent set is open.
+* `is_closed`: the spectrum is closed.
+* `subset_closed_ball_norm`: the spectrum is a subset of closed disk of radius equal to the norm.
+* `is_compact`: the spectrum is compact.
+* `spectral_radius_le_norm`: the spectral radius is bounded above by the norm.
+* `resolvent_has_deriv_at`: the resolvent function is differentiable on the resolvent set.
+
+
+## TODO
+
+* after we have Liouville's theorem, prove that the spectrum is nonempty when the
+  scalar field is ℂ.
+* compute all derivatives of `resolvent a`.
+
+-/
+
+/-- The *spectral radius* is the supremum of the norm of elements in the spectrum.
+    The spectrum is compact, and so when it is nonempty, this supremum can be
+    achieved. Note that it is possible for `spectrum 𝕜 a = ∅`.  In this case,
+    `spectral_radius a = 0` by the definition of `Sup` on `ℝ`.-/
+noncomputable def spectral_radius (𝕜 : Type*) {A : Type*} [normed_field 𝕜] [ring A]
+  [algebra 𝕜 A] (a : A) : ℝ :=
+Sup (norm '' (spectrum 𝕜 a))
+
+
+namespace spectrum
+
+section spectrum_compact
+
+variables {𝕜 : Type*} {A : Type*}
+variables [normed_field 𝕜] [normed_ring A] [normed_algebra 𝕜 A] [complete_space A]
+
+local notation `σ` := spectrum 𝕜
+local notation `ρ` := resolvent_set 𝕜
+local notation `↑ₐ` := algebra_map 𝕜 A
+
+lemma mem_resolvent_set_of_nearby {a : A} {k k' : 𝕜} (hk : k ∈ ρ a)
+  (hkk' : ∥k' - k∥ < ∥(↑hk.unit⁻¹ : A)∥⁻¹) :
+  k' ∈ ρ a :=
+begin
+  refine (units.unit_of_nearby hk.unit (↑ₐk' - a) _).is_unit,
+  calc ∥(↑ₐk' - a) - (↑ₐk - a)∥
+       = ∥↑ₐ(k' - k)∥         : by rw [ring_hom.map_sub, sub_sub_sub_cancel_right]
+  ...  = ∥k' - k∥ * ∥(1 : A)∥ : by rw [algebra.algebra_map_eq_smul_one,norm_smul]
+  ...  = ∥k' - k∥             : by simp [normed_algebra.norm_one 𝕜 A]
+  ...  < ∥↑hk.unit⁻¹∥⁻¹       : hkk',
+end
+
+lemma is_open_resolvent_set (a : A) : is_open (ρ a) :=
+begin
+  haveI := normed_algebra.nontrivial 𝕜 A,
+  apply metric.is_open_iff.mpr,
+  intros k hk,
+  refine ⟨∥↑hk.unit⁻¹∥⁻¹, inv_pos.mpr (units.norm_pos (hk.unit⁻¹)), _⟩,
+  intros k' hk',
+  rw [metric.mem_ball, dist_eq_norm] at hk',
+  exact mem_resolvent_set_of_nearby hk hk',
+end
+
+/-- The `resolvent_set` as a term of `opens 𝕜` -/
+def open_resolvent_set (a : A) : topological_space.opens 𝕜 :=
+⟨ρ a, is_open_resolvent_set a⟩
```

Comment:
I think you can omit this one.  We usually use `is_open`; the bundled version `topological_space.opens` is only used in a few parts of the library (I think in sheaf theory, for example) where one needs a name for the whole collection.

Reply:
okay, we probably will want the corresponding type for the spectrum though (nonempty compact) because then we can say that the spectrum is an upper semicontinuous function w.r.t. the Hausdorff metric. But we're not there yet.

Reply:
Not sure I follow -- for me, upper semicontinuity makes sense for functions into an ordered type, and "the type of compact subsets of C, equipped with the Hausdorff metric" does not have a natural order?  But indeed we're not there yet :)